src/SemiFungible1155.sol

```diff
@@ -14,27 +15,43 @@ error NotAllowed();
 error TypeMismatch();
 error FractionalBurn();
 
+/// @title Contract for minting semi-fungible EIP1155 tokens
+/// @author bitbeckers
+/// @notice Extends { Upgradeable1155 } token with semi-fungible properties and the concept of `units`
+/// @dev Adds split bit strategy as described in [EIP-1155](https://eips.ethereum.org/EIPS/eip-1155#non-fungible-tokens)
 contract SemiFungible1155 is Upgradeable1155 {
+    /// @dev Counter used to generate next typeID.
     uint256 public typeCounter;
-    // Use a split bit implementation.
-    // Store the type in the upper 128 bits..
+
+    /// @dev Bitmask used to expose only upper 128 bits of uint256
```

Comment:
How many bits we spend on the baseId vs the index is related to how many hypercerts do we expect to exist, and how many fractions of a hypercert do we expect to exist?

Just to brainstorm, is there any way to change this in an upgrade? Maybe the answer is no practical way and maybe we're just okay with that

Reply:
We'd break the accounting of previously minted tokens because the index has been using an uint128 thus far and shifting that would break the indexes. We could choose to for instance use uint192 + uint64. As we are only minting registrations, I think you could perfectly solve this with a new deployment using different balances.

Reply:
` I think you could perfectly solve this with a new deployment using different balances.`

Could you elaborate what this means?

Reply:
With solving I meant changing the available bits for upper/lower. In other words, don't change the contract but deploy a new one

Reply:
Oh I see. Yes, but there'd be no way to migrate state between the contracts, so we'd just start tracking a second contract correct?